File: python-lambda/lambda/handlers/load_user.py
import json
import logging
import os

import boto3

logger = logging.getLogger(__name__)

dynamo = boto3.resource('dynamodb')
TABLE_NAME = os.environ['USERS_TABLE_NAME']


def load_user(event: dict) -> dict:
    user_id = (event.get('pathParameters') or {}).get('userId')
    if not user_id:
        return _json(400, {'error': 'userId path parameter is required'})

    logger.info(f"Loading user with id: {user_id}")

    try:
        result = dynamo.Table(TABLE_NAME).get_item(Key={'userId': user_id})
    except Exception:
        logger.error('Failed to load user', exc_info=True)
        return _json(500, {'error': 'Internal server error'})

    item = result.get('Item')
    if not item:
        logger.error(f"User not found with id: {user_id}.")
        return _json(404, {'error': 'User not found'})

    logger.info(f"User loaded successfully with id: {user_id}.")
    return _json(200, item)


def _json(status_code: int, body: object) -> dict:
    return {
        'statusCode': status_code,
        'headers': {'Content-Type': 'application/json'},
        'body': json.dumps(body),
    }

File: java-lambda/lambda/src/main/java/com/trace0/javalambda/handlers/LoadUser.java
package com.trace0.javalambda.handlers;

import com.amazonaws.services.lambda.runtime.events.APIGatewayProxyRequestEvent;
import com.amazonaws.services.lambda.runtime.events.APIGatewayProxyResponseEvent;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.trace0.javalambda.Response;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import software.amazon.awssdk.services.dynamodb.DynamoDbClient;
import software.amazon.awssdk.services.dynamodb.model.AttributeValue;
import software.amazon.awssdk.services.dynamodb.model.GetItemRequest;
import software.amazon.awssdk.services.dynamodb.model.GetItemResponse;

import java.util.Map;

public class LoadUser {

    private static final Logger logger = LogManager.getLogger(LoadUser.class);
    private static final DynamoDbClient dynamo = DynamoDbClient.create();
    private static final String TABLE_NAME = System.getenv("USERS_TABLE_NAME");
    private static final ObjectMapper mapper = new ObjectMapper();

    public APIGatewayProxyResponseEvent handle(APIGatewayProxyRequestEvent event) {
        Map<String, String> pathParams = event.getPathParameters();
        String userId = pathParams != null ? pathParams.get("userId") : null;
        if (userId == null) {
            return Response.json(400, "{\"error\":\"userId path parameter is required\"}");
        }

        logger.info("Loading user with id: {}", userId);

        GetItemResponse result;
        try {
            result = dynamo.getItem(GetItemRequest.builder()
                    .tableName(TABLE_NAME)
                    .key(Map.of("userId", AttributeValue.fromS(userId)))
                    .build());
        } catch (Exception e) {
            logger.error("Failed to load user", e);
            return Response.json(500, "{\"error\":\"Internal server error\"}");
        }

        if (!result.hasItem() || result.item().isEmpty()) {
            logger.error("User not found with id: {}", userId);
            return Response.json(404, "{\"error\":\"User not found\"}");
        }

        Map<String, AttributeValue> item = result.item();
        Map<String, String> user = Map.of(
                "userId", item.get("userId").s(),
                "name", item.get("name").s(),
                "email", item.get("email").s(),
                "createdAt", item.get("createdAt").s()
        );

        logger.info("User loaded successfully with id: {}", userId);
        try {
            return Response.json(200, mapper.writeValueAsString(user));
        } catch (Exception e) {
            return Response.json(500, "{\"error\":\"Internal server error\"}");
        }
    }
}
